### ISSUE TITLE
Support Parquet bloom filter

### ISSUE THREAD

**Issue by zhouyuan** [facebookincubator/velox#4414]
### Description

Hi Velox community, 

Parquet 1.12 has defined the bloom filter feature which allows to filter out more data precisely when reading. Here's the detail spec in parquet-format: https://github.com/apache/parquet-format/blob/master/BloomFilter.md 

Recent Spark has also imported this feature and will automatically do bloom filtering if it exists in metadata.
https://github.com/apache/spark/blob/branch-3.3/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala#L1895-L1923

It would be great to have this feature in Velox native parquet reader as it would benefit the table scan performance a lot. The write code path may require time as arrow parquet write seems does not support this feature yet. 

CC: @mbasmanova @oerling 

Thanks, -yuan

**Comment by mbasmanova** [comment]
@zhouyuan Would you be interested in adding this functionality?

**Comment by zhouyuan** [comment]
@mbasmanova yes, have done some local tests here and I could help to implement this. 

-yuan

**Comment by Yuhta** [comment]
Can you share your implementation plan here?  This is not Parquet-only work and need some common filter framework support.

How would you design the bloom filter interface that will be passed to filters?  How do you decide the hashing for strings?  This bloom filter interface must be generic enough and not restricted to the Parquet bloom filter implementation above.  It needs to handle both int64 and strings.  Once the interface is decided (say an interface class `IBloomFilter`), we can add `Filter::testBloomFilter(const IBloomFilter&) const` to all the filters.

Then for the Parquet specific part, you can implement a subclass of `IBloomFilter`, and in `ParquetData::rowGroupMatches`, parse the Parquet metadata into this subclass, then check if the filter is disjunctive with the bloom filter using `Filter::testBloomFilter` you added before.

**Comment by oerling** [comment]
There is a method in ScanSpec that checks a stats against a Filter. The Bloom filter itself is encapsulated inside ColumnStats. The hash  for different data types is specified by the file format, as is the layout of the Bloom filter. We switch on the format kind by subclassing ColumnStats, for example. 

 

This operates on metadata only. 

 

Pushing down an approximate filter, like an approxinmate variant of BigintValues is a different thing and not needed for this occasion.
